src/org/freedesktop/gstreamer/Pad.java

```diff
@@ -660,34 +715,42 @@ public boolean hasCurrentCaps() {
     private static class Handle extends GstObject.Handle {
         
         private final Set<NativeLong> probes;
-        
+
+        // The Pads are stored here, so that they are not GC'ed while the Probe still exists
+        @SuppressWarnings("MismatchedQueryAndUpdateOfCollection")
+        private final Map<NativeLong, Pad> probePads;
+
         private Handle(GstPadPtr ptr, boolean ownsHandle) {
             super(ptr, ownsHandle);
             probes = new HashSet<>();
+            probePads = new HashMap<>();
         }
 
         @Override
         protected GstPadPtr getPointer() {
             return (GstPadPtr) super.getPointer();
         }
-        
-        private synchronized NativeLong addProbe(int mask, GstPadAPI.PadProbeCallback probe) {
+
+        private synchronized NativeLong addProbe(int mask, GstPadAPI.PadProbeCallback probe, Pad pad) {
```

Comment:
Sorry, but we can't keep reference to `Pad` in `Pad.Handle`. This will cause memory leaks, even if it fixes your immediate problem.

Reply:
I had a Test-Program (see Gist in the original Issue) spin around .block() a couple million times and monitored the stack with VisualVM, and there is no memory-leak here.

The String Ref to the Pad is stored alongside with the Probe-Callback and removed when the Probe-Callback is removed. The GC is then free to clean the Pad up. The reference in Handle is only held as long as there is a Probe pending.

Reply:
I'm concerned about other methods than `block()`.

Reply:
As the [Probe-Callback](https://github.com/gstreamer-java/gst1-java-core/blob/master/src/org/freedesktop/gstreamer/Pad.java#L395) (not only for block but for all Probes) refers to the Pad, the Pad-Instance *has* to live as long as the Probe does. When the Probe goes away, it *has* to call `Handle.removeProbe` (it does so through `GCallback.disconnect` when the Callback is removed).

Thus I cannot see, how the Pad-Instance in the Handle could form a Memleak without the Probe also forming one. It should be fairly simple to generate a small piece of Code that exposes such a leak ant it should be fairly easy to spot in VisualVM.

Reply:
You mean because it's an inner class (not the method argument)?  That the Pad and Probe reference each other isn't a problem. But `Handle` should not under any circumstances have a hard reference to the object it's a handle for.  `Handle` has a very particular job to do around memory handling, and this risks bringing in huge problems.

Reply:
I see that I mis-used (and did not understand) Handle here. Let's talk about alternatives. I though about implementing the Probe-Object as a Closure that holds an implicit reference to the surrounding Pad; something like
```
Pad padRef = this;
final GstPadAPI.PadProbeCallback probe = new GstPadAPI.PadProbeCallback() {
   …
   return listener.eventReceived(padRef, event);
   …
}
```
This should also keep a Reference, as long as the inner Probe-Object exists. I have to check this, though.

Reply:
I tried that but it alone does not help. To copy from the Issue (and remind myself of the actual problem): The root cause is, that
 - the GstPadAPI.PadProbeCallback instance is only held on the Pad object
 - nothing guarantees that the Pad object lives long enough so that not Path, the Pad and the GstPadAPI.PadProbeCallback instance are GCed while the Probe is in place.

So somewhere the Pad *OR* the GstPadAPI.PadProbeCallback instance needs to be stored *within the Java Object-Graph* while the Probe is active *within the GStreamer Stack*.

*OR*, because we can attach a Reference to the Pad to the GstPadAPI.PadProbeCallback by making it a final reference as mentioned above.

Reply:
Handle is part of a dual hierarchy that is used in NativeObject.NativeRef - https://github.com/gstreamer-java/gst1-java-core/blob/master/src/org/freedesktop/gstreamer/glib/NativeObject.java#L241  Note that that is a weak reference to the native object with a strong reference to the handle - it's how the memory handling works and the handle having a strong reference back to the actual native object is likely to cause problems.

My inclination may still be to store the Pads inside the Element where necessary.  But let's clarify where exactly the Pad is coming from, and the ownership / refcounts on it?  I have some code I found that tests some probe calls for GC issues from earlier problems - not sure when - but that doesn't have an issue.

Reply:
The actual Code triggering the Problem reliably is this: https://github.com/MaZderMind/gintercom-mixing-core/blob/master/src/main/java/de/mazdermind/gintercom/mixingcore/Panel.java#L123-L132

As yo ucan see, the Pad-Object is generated from `Pad mixerPad = ((GhostPad) pad).getTarget();`

Reply:
Can you see if adding `@CallerOwnsReturn` to `gst_element_get_request_pad` helps?   eg. https://github.com/gstreamer-java/gst1-java-core/blob/master/src/org/freedesktop/gstreamer/lowlevel/GstElementAPI.java#L77 doesn't match up with `transfer-full` at https://developer.gnome.org/gstreamer/stable/GstElement.html#gst-element-get-request-pad